Terraform/atru/mcp server 020 (#580)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/docs/tools/mcp-server/deploy.mdx

# Deploy the Terraform MCP server

> The Terraform MCP server runs in a Docker container that you can deploy to either your local workstation or a remote instance. When running the server locally, start it in the default `stdio` mode. In this mode, the server communicates over standard input/output. When running the server on a remote instance, start the server in `streamable-http` mode so that you can communicate with the server using the Streamable HTTP transport.

Reviewer: ```suggestion
The Terraform MCP (Model Context Protocol) server runs in a Docker container that you can deploy to either your local workstation or a remote instance. When running the server locally, start it in the default `stdio` mode. In this mode, the server communicates over standard input/output. When running the server on a remote instance, start the server in `streamable-http` mode so that you can communicate with the server using the Streamable HTTP transport.
```

Define initialism on first use.

> The Terraform MCP server runs in a Docker container that you can deploy to either your local workstation or a remote instance. When running the server locally, start it in the default `stdio` mode. In this mode, the server receives and sends messages in JSON format over RPC. When running the server on a remote instance, start the server in `streamable-http` mode so that you can communicate with the server using the Streamable HTTP transport.

Reviewer 2: stdio communicates over standard input/output

> The Terraform MCP server runs in a Docker container that you can deploy to either your local workstation or a remote instance. When running the server locally, start it in the default `stdio` mode. In this mode, the server receives and sends messages in JSON format over RPC. When running the server on a remote instance, start the server in `streamable-http` mode so that you can communicate with the server using the Streamable HTTP transport.

Author: ```suggestion
The Terraform MCP server runs in a Docker container that you can deploy to either your local workstation or a remote instance. When running the server locally, start it in the default `stdio` mode. In this mode, the server communicates over standard input/output. When running the server on a remote instance, start the server in `streamable-http` mode so that you can communicate with the server using the Streamable HTTP transport.
```

## Install the server

You can install the server using the following methods if you intend to run the server locally:

> - Install compiled the binary into your client. The binaries are available at [releases.hashicorp.com](https://releases.hashicorp.com/terraform-mcp-server/).

Reviewer: ```suggestion
- Install the binary into your client. The binaries are available at [releases.hashicorp.com](https://releases.hashicorp.com/terraform-mcp-server/).
```

- Install from source into your client.

> For remote installations, download the source to the remote and build the Docker image.

Reviewer: Just curious, why isn't using a pre-compiled binary an option when running remotely?

For remote installations, download the source to the remote and build the Docker image.

### Compiled binary

Refer to [Requirements](#requirements) for supported clients.

<Tabs>

<Tab heading="VS Code">

To install the server so that it's available for all workspaces, add the following JSON block to your user setting JSON file:

```json
{
"servers": {

>    "terraform-mcp-server": {

Reviewer: Can we tell them where this file is located?

        "command": "docker",
        "args": [
          "run",
          "-i",
          "--rm",
          "hashicorp/terraform-mcp-server"
        ]
      }
    }
  }
}
```

Alternatively, you can install the server in specific workspaces. Create an `mcp.json` file in the `.vscode` directory of your workspace and add the following JSON block:

```json
{
  "servers": {
    "terraform": {
      "command": "docker",
      "args": [
        "run",
        "-i",
        "--rm",
        "hashicorp/terraform-mcp-server"
      ]
    }
  }
}
```

</Tab>

<Tab heading="Claude Desktop, Amazon Q Developer and CLI">

Create an `mcp.json` file and add the following configuration:
 
```json
{
"mcpServers": {
  "servers": {
   "hcp-terraform": {
     "command": "docker",
     "args": [
        "run",
        "-i",
        "--rm",

>         "hashicorp/terraform-mcp-server:0.1.0"

Reviewer 2: Feel free to remove version from here too

Author: ```suggestion
        "hashicorp/terraform-mcp-server:<version>"
```

       ]
     }
   }
 }
}

```

</Tab>

</Tabs>

Start or restart Docker to start the container using the standard transport.

### Install from source

> You can pull the source from either the `latest` release version or pull from the `main` branch to install from source.

Reviewer: Can/should we give some guidance around which to choose? Should we mention installing from a release tag as well?

For automation contexts, it's usually better to use a specific release to avoid unintended changes breaking automation.

Latest:

```shell-session
$ go install github.com/hashicorp/terraform-mcp-server/cmd/terraform-mcp-server@latest
```

Main:

```shell-session

> $ go install go install github.com/hashicorp/terraform-mcp-server/cmd/terraform-mcp-server@main

Reviewer: ```suggestion
$ go install github.com/hashicorp/terraform-mcp-server/cmd/terraform-mcp-server@main
```

```

After pulling the image, add the following JSON block to your client and specify the path to the server binary in the `command` attribute:

```json
{
  "mcp": {
    "servers": {
      "terraform": {
        "command": "/path/to/terraform-mcp-server",
        "args": ["stdio"]
      }
    }
  }
}
```

Start or restart Docker to start the container using the standard transport.

## Configure environment variables

If you intend to run the MCP server on a remote instance, log into the remote machine and specify the following environment variables:

| Variable | Description | Default |
|----------|-------------|---------|
| `TRANSPORT_MODE` | Set to `streamable-http` to enable HTTP transport | `stdio` |
| `TRANSPORT_HOST` | Host to bind the HTTP server | `0.0.0.0` |
| `TRANSPORT_PORT` | HTTP server port | `8080` |

## Build the Docker image 

You can download the source to your local machine or to a remote instance and build the server locally. 

1. Clone the repository:

```bash
git clone https://github.com/hashicorp/terraform-mcp-server.git
cd terraform-mcp-server
```

> 

Reviewer: Similar to above - they could switch to a tag to build that specific version.
